### ISSUE TITLE
Support for SHA256

### ISSUE THREAD

**Issue by pjbgf** [go-git/go-git#706]
Implement SHA256 support as per https://git-scm.com/docs/hash-function-transition/.

High level changes:
- [ ] Repository Format Extension
- [ ] Loose object index
- [ ] Bidirectional (SHA-1 <-> SHA-256) mapping table
- [ ] Fetching from SHA1 server
- [ ] Signed Commits
- [ ] Signed Tags
- [ ] Mergetag
- [ ] Submodules
- [ ] Shallow clone (unsupported - fail gracefully)

**Comment by AdamMajer** [comment]
Thanks for working on this. I think the transition doesn't yet need to be implemented, since it doesn't really work in git tools anyway. What should be implemented is simply ability to work with SHA256 repositories - I think this is the scope of the work here.

SHA1 <> SHA256 repo interop probably will need some more thought and implementation upstream first.

**Comment by AdamMajer** [comment]
Seems a duplicate of https://github.com/go-git/go-git/issues/229 although more work is done here so maybe close the other issue.

**Comment by pjbgf** [comment]
@AdamMajer thanks for calling out the duplicated issue - just closed it.

I agree with your comment on the transition. Given that `go-git` aims to be aligned with upstream, the items pending for SHA256 will always have a hard blocker in terms of whether or not upstream already merged the given feature/change.

#707 Already added support to create and make commits in SHA256 repositories, but this is still experimental - hence why it is blocked by a build tag.

**Comment by AdamMajer** [comment]
@pjbgf , just to make sure I'm reading the patch you linked correctly, this basically just allows go-git to use only sha256 or sha1, depending on the build tag?  There is no possibility of using both hash types in one program?

**Comment by pjbgf** [comment]
@AdamMajer that's correct. That is a temporary limitation which aims to keep backwards compatibility with the API - as `plumbing.Hash` defines the hash's length. There are also some performance/memory footprint considerations that will need to be considered when we implement that.

Once the bilateral mapping table is implemented, we should remove that limitation.

**Comment by gecube** [comment]
Hi!

I am waiting for changes, as I just experienced that Gitlab.com offers the new SHA-256 algorithm, but unfortunately tools bases on go-git could not work with such a repos...

![image](https://github.com/user-attachments/assets/bb8e7314-cd6c-43ca-9e67-61a26c125c43)

**Comment by pjbgf** [comment]
@gecube thanks for the ping - I wasn't aware of that support from Gitlab. :bow: 
The `packfile` refactoring is well under progress, hopefully by next month I will be able to get that merged against `v6-exp`.

**Comment by bk2204** [comment]
I'll just mention that Git 3.0 will make SHA-256 repositories (and reftable) the default and that the tentative plan for release is about a year from now (although that could change).  It looks like this is already well underway here, though, which is encouraging to see.

**Comment by pjbgf** [comment]
@bk2204 thanks for the heads-up - and for the great presentation at Git Merge 2025. 👏
Hopefully we will get the core of these changes merged in the coming months so that way we have plenty of time for testing. 🤞

**Comment by pjbgf** [comment]
A [discussion](https://github.com/go-git/go-git/pull/1825#discussion_r2732245549) to refer to when implementing the bi-directional mapping table (a.k.a [translation table](https://git-scm.com/docs/hash-function-transition#_translation_table)).

**Comment by mcepl** [comment]
> A [discussion](https://github.com/go-git/go-git/pull/1825#discussion_r2732245549) to refer to when implementing the bi-directional mapping table (a.k.a [translation table](https://git-scm.com/docs/hash-function-transition#_translation_table)).

With reference to [the above](https://github.com/go-git/go-git/issues/706#issuecomment-1512999410), isn't the translation out of the scope of this issue?

**Comment by pjbgf** [comment]
@mcepl initially I was going to cover both sha256 and the compat extension as a single issue. But I agree that may be confusing for some users. I created a new issue just for the `compactObjectFormat` and have the items related to that from the description of this issue.

My understanding is that sha256 is already working, however I will close this issue once I finished refactoring the tests and ensure we have good enough coverage for it.

**Comment by mcepl** [comment]
I came here from Sourcehut, which is waiting for this bug to be resolved before they can start working on SHA256 support.

**Comment by mischief** [comment]
this would be great to get fixed, since this also affects the gitea 'tea' cli. gitea itself supports hosting sha256 repos, but 'tea' which uses go-git chokes on it.

**Comment by mischief** [comment]
also worth noting that the upcoming git 3.0 documents a default sha256 in https://github.com/git/git/blob/ab776a62a78576513ee121424adb19597fbb7613/Documentation/BreakingChanges.adoc?plain=1#L89.
